File: 890-find-and-replace-pattern/890-find-and-replace-pattern.cpp
class Solution {
public:
    bool isMatching(string word, string pattern)
    {
        unordered_map<char,char> mp;
        set<char> st;
        for(int i =0;i<pattern.size();i++)
        {
            if(mp.find(pattern[i]) != mp.end()){
                //pattern matching is already there:
                if(mp[pattern[i]] != word[i])
                    return false;
            }
            else{
                // new pattern matching is found:
                if(st.find(word[i]) != st.end())
                    return false;
                mp[pattern[i]] = word[i];
                st.insert(word[i]);
            }
        }
        return true;
    }
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> res;
        for(auto word:words){
            if(isMatching(word,pattern))
                res.push_back(word);
        }
        return res;
    }
};

File: 890-find-and-replace-pattern/890-find-and-replace-pattern.java
class Solution {
    public static boolean isMatching(String word, String pattern)
    {
        HashMap<Character,Character> mp = new HashMap<>();
        HashSet<Character>st = new HashSet<>();
        
        for(int i = 0;i<pattern.length();i++)
        {
            char wch = word.charAt(i);
            char pch = pattern.charAt(i);
            
            if(mp.containsKey(pch) == true){
                // pattern character is already mapped:
                if(mp.get(pch) != wch)
                    return false;
            }
            else{
                // pattern character is new for mapping:
                if(st.contains(wch))
                    return false;
                mp.put(pch,wch);
                st.add(wch);
            }
        }
        return true;
    }
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> res = new ArrayList<>();
        for(String word: words){
            if(isMatching(word,pattern))
                res.add(word);
        }
        return res;
    }
}